Lib/test/test_compileall.py

```diff
@@ -360,6 +361,234 @@ def test_ignore_symlink_destination(self):
+            compileall.compile_dir(self.directory, quiet=True, optimize=0,
+                                   hardlink_dupes=True)
+
+    def test_hardlink_deduplication_same_bytecode_all_opt(self):
+        # 'a = 0' produces the same bytecode for all optimization levels
+        path = os.path.join(self.directory, "test", "same_all")
+        os.makedirs(path)
+
+        simple_script = script_helper.make_script(path, "test_same_bytecode",
+                                                  "a = 0")
+        pyc_opt0 = importlib.util.cache_from_source(simple_script)
+        pyc_opt1 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=1)
+        pyc_opt2 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=2)
+
+        compileall.compile_dir(path, quiet=True, optimize=[0, 1, 2],
+                               hardlink_dupes=True)
+
+        # import pdb; pdb.set_trace()
+
+        # All three files should have the same inode (hardlinks)
+        self.assertEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt1).st_ino)
+        self.assertEqual(os.stat(pyc_opt1).st_ino, os.stat(pyc_opt2).st_ino)
+
+        for pyc_file in {pyc_opt0, pyc_opt1, pyc_opt2}:
+            os.unlink(pyc_file)
+
+        compileall.compile_dir(path, quiet=True, optimize=[0, 1, 2],
+                               hardlink_dupes=False)
+
+        # Deduplication disabled, all pyc files should have different inodes
+        self.assertNotEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt1).st_ino)
+        self.assertNotEqual(os.stat(pyc_opt1).st_ino, os.stat(pyc_opt2).st_ino)
+
+    def test_hardlink_deduplication_same_bytecode_some_opt(self):
+        # 'a = 0' produces the same bytecode for all optimization levels
+        # only two levels of optimization [0, 1] tested
+        path = os.path.join(self.directory, "test", "same_some")
+        os.makedirs(path)
+
+        simple_script = script_helper.make_script(path, "test_same_bytecode",
+                                                  "a = 0")
+        pyc_opt0 = importlib.util.cache_from_source(simple_script)
+        pyc_opt2 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=2)
+
+        compileall.compile_dir(path, quiet=True, optimize=[0, 2],
+                               hardlink_dupes=True)
+
+        # Both files should have the same inode (hardlink)
+        self.assertEqual(os.stat(pyc_opt0).st_ino,  os.stat(pyc_opt2).st_ino)
+
+        for pyc_file in {pyc_opt0, pyc_opt2}:
+            os.unlink(pyc_file)
+
+        compileall.compile_dir(path, quiet=True, force=True, optimize=[0, 2],
+                               hardlink_dupes=False)
+
+        # Deduplication disabled, both pyc files should have different inodes
+        self.assertNotEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt2).st_ino)
```

Comment:
This code is very close to test_hardlink_deduplication_same_bytecode_all_opt(). Can you try to write an helper function to factorize the code?

Same remark for tests below.

Reply:
I've tried to parametrize the three tests for the same bytecode using subTest. What do you think? Again, it's done only for unit tests because we might remove some CLI tests later.